### PR TITLE
#526 Preventing SSRF due to external resource refs in SVGs

### PR DESCRIPTION
**Pull request by actinium15**

### CODE REVIEW

**Comment by actinium15** [comment]
hi @haraldk, may I request you to provide guidance on the changeset? I'd be happy to work on improving it should you have any suggestions... Thanks!

**Comment by haraldk** [comment]
Sorry, haven't had the time to look into it yet. I will! 😀

-- 
Harald K

File: imageio/imageio-batik/src/main/java/com/twelvemonkeys/imageio/plugins/svg/SVGImageReader.java [hunk 1]
```diff
@@ -80,6 +80,7 @@
  */
 public class SVGImageReader extends ImageReaderBase {
     private Rasterizer rasterizer;
+    private boolean allowExternalResources;
```
Comment (haraldk) [inline imageio/imageio-batik/src/main/java/com/twelvemonkeys/imageio/plugins/svg/SVGImageReader.java]: Just thought about it, and we could do something like:

```suggestion
    private boolean allowExternalResources = "true".equalsIgnoreCase(System.getProperty("com.twelvemonkeys.imageio.plugins.svg.allowexternalresources"));
```

**Comment by haraldk** [comment]
Thanks, I was about to write that we could do this much simpler, but you beat me to it. 😀

The new version is nearly there. However I have some changes I would like to do:

1. Instead of the constant for the system property name, I would like to have a *single* constant for the *global setting* (ie: `final static boolean DEFAULT_ALLOW_EXTERNAL_RESOURCES = "true".equalsIgnoreCase(System.getProperty(...));` (as we now read the system property only once, we no longer need a constant for its name, just keep the string inline). This will fix the setting to the *initial* value of the system property, to avoid confusing behaviour if someone tries to change it at runtime. This is also more secure.
2. Unfortunately, this also affects our tests. But in reality, this is also a good thing, because test may run in parallell, and relying on setting/clearing system properties could potentially result in flaky tests. The tests that relies on external resources will have to be read using param.
3. Both the local settings (reader and param) will be initialized to the value of the global setting (as now, just the constant instead of the system property). 

With these changes, I hope to be able to merge your PR. 😀

Great work so far!

-- 
Harald K

**Comment by actinium15** [comment]
Thanks for your guidance @haraldk!

I've pushed another commit based on your latest review (I hope I got all the things you wanted right). Please do let me know in case I've missed something. Thanks for your understanding and apologies in case I've changed something the way you didn't expect (esp the part where I enabled the system-property for whole of the test-JVM in the pom - per my understanding that's the only way to be certain for reliable parallel execution of tests, though that did mean I could no longer assert the default (no system-property, no API invocation) behaviour in the test)

**Pull request merged**